### ISSUE TITLE
Warn when undentation of operators causes ambiguity when following a match or if-else block

### ISSUE THREAD

**Issue by abelbraaksma** [fsharp/fslang-suggestions#806]
# Warn for undentation ambiguities caused by operators

Note: this is _undentation_, or _indentation-relaxation_, not _indentation_.

I propose we raise a warning when undentation relaxation for operators will lead to ambiguities. This warning can be raised when the parser detects that after undenting an operator for the width of the operator, would make it belong to the previous block instead of the last statement. The warning could be:

> Warning FS9999: Operator undentation ambiguity detected for operator "<??>" causes it to belong to the previous line, instead of the previous block. To avoid this warning, indent the operator to the same indentation level as the previous line, or outdent it to have it belong to the block instead.

**Rationale**

It is well established that F# is a whitespace-sensitive language, where generally related statements must start at the same indentation level.

For operators, there's an exception, where the parser will allow the operator to be _undented_ until the rh-side of the operator is at the same line as the previous line. When the operator changes type, this is not a problem and wrongful use will lead to errors:

```f#
// this example works
let f v =
   match v with
   | None -> []
   | Some i -> [1..i]
   |> Seq.sum

// this example errors:
let f v =
   match v with
   | None ->
      []
   | Some i -> 
      [1..i]
   |> Seq.sum    // caused by `Seq.sum` starting on the same level as `[1..i]`
                 // error: The type ''a list' does not match the type 'int'
```

In the examples above, it is clear, because the type inference will raise an exception (though the exception by itself will be hard to understand as it will be about the type, not the undentation error).

If we change this a little and create an operator for logging, it becomes must harder to detect:

```f#
// this example uses the default of 4 spaces for indentation
let (<%>) x y = Console.WriteLine y; x
let f v =
    match v with
    | None -> []
    | Some i -> 
        [1..i]
   <%> "logging"    // only hit when match is Some, never hits for None

// it becomes even harder to spot in an if-else statement
// operators like these are common in FParsec and other libraries
let (<??>) x y = Console.WriteLine y; x
let f v =
    if condition then []
    else [1..i]
    <%%> "Tested condition"
```

The problem is most apparent with operators that don't change type, because otherwise the chance of detection is much larger as it will likely raise a compile-time error.

The existing way of approaching this problem in F# is: to live with it and spot such errors visually (and as multiple reports have shown, this can be rather hard, even for seasoned developers).

## Pros and Cons 

The advantages are:

* an improved coding experience and less extremely-hard-to-find bugs. 
* F# is a very clear language and it's strength lies in the fact that type inference and strong, strict typing rules lead to writing correct code very easily and with little debugging. This improves on that experience.
* A warning allows the programmer to switch it off, or to improve his coding style for readability.

## Extra information

Estimated cost (XS, S, M, L, XL, XXL): XS (I really hope that's true)

Related suggestions: 

* This has been discussed before on StackOverflow (I'll update when I find it back)
* And here: https://github.com/dotnet/fsharp/issues/6136
* And ... (I'll update, I've seen several reports in the past, or add them in the comments)

## Affidavit (please submit!)

Please tick this by placing a cross in the box:
* [x] This is not a question (e.g. like one you might ask on [stackoverflow](http://stackoverflow.com)) and I have searched stackoverflow for discussions of this issue
* [x] I have [searched both open and closed suggestions on this site](http://github.com/fsharp/fslang-suggestions/issues) and believe this is not a duplicate
* [x] This is not something which has obviously "already been decided" in previous versions of F#.  If you're questioning a fundamental design decision that has obviously already been taken (e.g. "Make F# untyped") then please don't submit it.

Please tick all that apply:
* [x] This is not a breaking change to the F# language design
* [x] I or my company would be willing to help implement and/or test this

If this is going to be accepted and someone can point me to the location where the indentation-relaxation of operators takes place, I'll gladly take a look to see if I can have that code raise a warning for ambiguous scenarios.

**Comment by abelbraaksma** [comment]
To quote @dsyme [from this above referenced thread](https://github.com/dotnet/fsharp/issues/1019#issuecomment-205070141) of 2016, a bug report by @smoothdeveloper (thanks for pointing it out to me):

> @smoothdeveloper Please consider opening a suggestion on http://fslang.uservoice.com to request a warning for the first of these cases. It would be great if you could formulate an exact proposed rule (e.g. one that would cover the interaction with "match", but no more, unless you identify other similar interactions).

To answer that last line, I think we can do one of two things:

* special-case branch instructions where this is relevant (`if .. then .. else`, `match` and `match!`)
* or generalize the case where the parser does a test each time it encounters an operator. The test should be something like: _upon encountering a left-most operator, that starts at the same col position as the previous line, if I erase the operator and the trailing whitespace, does this line then still belong to the same previous block?_

Not sure whether either of above options is easy or hard to detect in practice and whether corner cases exist that are not covered by such a simple rule.

**Comment by charlesroddie** [comment]
> I propose we raise a warning when undentation relaxation for operators will lead to ambiguities.

These undentations are bad in general surely. That's what the ambiguities show. Rather than warning only when bad undentations cause ambiguities, we should warn whenever they are made. That would be a lot clearer to users and also simpler for the compiler.

I.e. in
```
<spaces> <code1>
<spaces> <operator> <code2>
```
`operator` (and not just `code2`) should align at least with `code1`.

**Comment by abelbraaksma** [comment]
> I don't see why though. Is this considered good style?

@charlesroddie I don't think anybody considers it good style. At best it's mentioned in some text books. But in practice, I see everyone align on the start of the operator.

The only exceptions to the rule are, however, any form of parens, brackets, curlies. Here, the first item starting with an opening paren, bracket or curly, is _undented_ one character, so that the actual code inside the brackets aligns.

Most bracket constructs use one character, but there's also `[|` and `{|`. I agree it's a good idea to warn _always_, but than with the exception of all kinds of bracket.

**Comment by charlesroddie** [comment]
Shouldn't the scope of this issue be undentation of operators, since bracket undentation doesn't give the ambiguities mentioned?

**Comment by abelbraaksma** [comment]
Thanks for approving this, @dsyme. I'll see if I am create an RFC.

**Comment by goswinr** [comment]
Would the suggested Warning show here too?
https://stackoverflow.com/questions/64784154/indentation-change-after-if-else-expression-not-taken-into-account

**Comment by abelbraaksma** [comment]
@goswinr, yes, this suggestion was meant specifically for such cases, see my description in the original issue, as that describes your SO question exactly.

**Comment by goswinr** [comment]
I really hope this gets into the language! It has bitten me more than once. Especially when I write a quick script for some side effects logic here or there. (Incomplete IF expressions that return UNIT, like in my SO question)  It’s hard to debug. And I couldn’t even plausibly explain the reason why its like that to an F# beginner. Someone on SO even answered  its a bug and i should raise an issue. He later deleted that answer.
